NHibernate StaleObjectStateException upgrade guidance improvements (#4750)
Particular/docs.particular.net: persistence/upgrades/nhibernate-8to8.1.md

## Upgrade paths

NHibernate 5 defaults to `DateTime2(7)` type, all newly created saga tables will use that new type for storing .NET `DateTime` values. These new sagas will work correctly with Version 8.1 of NHibernate persistence.

All existing sagas that use `DateTime` type in their tables stop working with Version 8.1 of NHibernate persistence due to the optimistic concurrency problem manifesting in `NHibernate.StaleObjectStateException` being thrown.

There are three upgrade paths to make these existing sagas work.

1. Alter schema to use the `DateTime2(7)` type
2. Use explicit version column for Optimistic Concurrency Control
3. Enforce NHibernate 4 backward compatibility mode
4. Set SQL compatibility mode to SQL Server 2014

INFO: For saga tables it is advised to both **upgrade the database schemas to use `DateTime2(7)`** and to **use an explicit version column**.

### Alter schema to use the `DateTime2(7)` type

> Manually alter the schema of existing saga tables to use the new `DateTime2(7)` data type. Such alteration can be done in-place without any data/precision loss.

Author: ```suggestion
Manually alter the schema of existing saga tables to use the new `DateTime2(7)` data type. Such alteration can be done in-place without any data/precision loss.

```sql
ALTER TABLE [MySagaTable] ALTER COLUMN [OneOfMyColumns] DATETIME2 NULL
```
```

NHibernate does not alter columns manually.

### Use explicit version column for Optimistic Concurrency Control

> NHibernate persistence has an option to use [explicit version column](/persistence/nhibernate/saga-concurrency.md#custom-behavior-explicit-version) for optimistic concurrency checks (OCC).

Author: ```suggestion
NHibernate persistence has an option to use an [explicit version column](/persistence/nhibernate/saga-concurrency.md#custom-behavior-explicit-version) for optimistic concurrency checks (OCC) and during an update or delete only a single `int` column will be compared.
```

> This requires changing the sagas code and altering the saga table schema by adding a column.

Author: ```suggestion
This requires adding a version property to the code of sagas and altering the database saga table schema by adding a column.
```
